3단계 - JDBC 라이브러리 구현

willing1026 opened the pull request:
안녕하세요 피드백 부탁드립니다! 🙇 

1. step2 에 대한 피드백 반영
   * BasicRowMapper를 만들어서 RowMapper에 대한 공통기능 관리
   * UserRowMapper 기능 간소화 (개발자가 적은일을 하도록 변경)
   * 그 외 피드백 반영

2. custom Exception 추가

willing1026 committed:
# modify : 타입추론을 위해 매개변수 RowMapper에 타입 명시 ( RowMapper<T> )

willing1026 committed:
# modify : jdbcTemplate 재사용하도록 구조 개선

willing1026 committed:
# modify : User Update 방식 개선 ( UserUpdatedDto, User 객체 의존 분리 )

- DataBase : 변경된 user객체 update method 추가
- UserApiCOntroller : update 기능수정
- UserUpdatedDto: 불필요한 기능 제거

willing1026 committed:
# modify & delete : UserRowMapper, UserListRowMapper 중복제거

##  modify
- RowMapper : return type 변경 ( T -> List<T> )
- UserRowMapper : 변경된 return type 적용
- JdbcTemplate : rowMapper 변경에 따른 selectOne 기능 수정
- UserDao : 사용하는 mapper 정리

## delete
- UserListRowMapper : 중복. 삭제

willing1026 committed:
# add & modify : rowMapper 공통기능 분리. BasicRowMapper 생성

## add
- BasicRowMapper : 도메인 클래스와 매핑할 컬럼명들을 넘기면 각 컬럼별로 알맞는 타입의 데이터를 세팅하는 기능 구현

## modify
- UserDao, RowMapper, UserRowMapper, JdbcTemplate : rowMapper 변경에 따른 수정

willing1026 committed:
# add & modify : custome Exception 추가

## add
- DataAccessException : SQLException 등 jdbc 기능관련 오류 처리

## modify
- UserDao, BasiRowMapper : 발생 에러 catch, 예외처리

kingbbode commented on src/main/java/next/dao/mapper/row/RowMapper.java:
```diff
@@ -2,7 +2,8 @@
 
 public interface RowMapper<T> {
-    public T mapResult(ResultSet rs) throws SQLException;
+    public List<T> mapResult(ResultSet rs) throws SQLException, IllegalAccessException;
```
BasicRowMapper 의 등장으로 사용자가 사용하기 더 어려워진 것은 아닐까요? BasicRowMapper 와 RowMapper 에 대한 연결점에 대한 부분과 사용성 또한 모호하다고 생각이 됩니다.

지난번 제 피드백 중 리스트 처리에 대한 중복 처리를 위해 이러한 구조가 나왔을까요?
중복에 대해서 피드백을 드렸던 이유는 RowMapper 의 의미가 모호했기 때문이였어요. 우리는 데이터베이스의 결과들을 Table이라고 이야기하고 그 안에 단 건의 데이터를 Row 라고 말합니다. 즉 RowMapper 는 단건의 데이터를 뜻하는 Row를 어떻게 처리할지만 정의되어야 한다는 의미입니다.

단건을 처리하는 RowMapper 와 복수건을 처리하는 RowMapper 는 다를 필요가 없지 않을까요? 결과를 순회하여 리스트로 반환할 것인가, 단건에 대한 처리만 할 것인가는 라이브러리의 몫이겠지요 :)

kingbbode commented on src/main/java/next/dao/mapper/row/BasicRowMapper.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.*;
+import java.util.stream.Stream;
+
+public class BasicRowMapper<T> {
```
사용자에게 구현체에 대한 특별한 방식의 사용을 요구하는 것이 어려운 것 같아요 :)

kingbbode reviewed (changes requested):
의견 남깁니다 :)
다시 한번 부탁드려요.

willing1026 committed:
# modify : 라이브러리가 RowMapper 구현체 제공하도록 개선

- RowMapper : list가 아닌  한 건 (row) 를  mapping 하도록 수정
- BasicRowMapper : UserRowMapper 제거 및 BasicRowMapper를 공통으로 사용할 수 있게 변경
- RowMapperFactory : 사용자에게 RowMapper 구현체 제공
- JdbcTemplate : RowMapper에 존재하던 selectOne, selectAll에 대한 로직을  여기서 구현
- UserDao : RowMapperFactory를 이용해서 rowMapper 구현체 만들도록 변경

willing1026 commented:
@kingbbode 
안녕하세요! DM으로 알려주신 내용대로 만들지는 못했는데.. 아래와 같이 수정했습니다. 
구현체라는 의미를 잘못 이해했을수도 있는데 `구현체=생성된객체`라고 이해하고 수정했습니다.
부족하지만 피드백 부탁드립니다 🙇 

- RowMapper 의미에 맞게 List가 아닌 하나의 Row를 처리하도록 변경
- RowMapperFactory를 만들고, 사용자에게 rowMapper구현체(BasicRowMapper)를 생성하여 반환
- JdbcTemplate의 `selectOne`, `selectAll`에 따라 row를 담는 방식을 알맞게 변경

kingbbode commented on src/main/java/next/dao/mapper/row/BasicRowMapper.java:
```diff
@@ -0,0 +1,72 @@
+    private Map<String, Field> fields;
+    private List<String> columns;
+
+    public BasicRowMapper(Class<T> clazz, String... columns) {
```
clazz 만으로도 필드이름을 통해 결과를 만들어낼 수 있지 않을까요? columns를 함께 입력받는 의도가 무엇인가요?

처음 이 기능을 제안드린 것은 RowMapper와 도메인 클래스 간의 일관성을 클래스 기준으로 보장해주어 사용자에게 편리함을 주기 위함이였어요 :)

willing1026 replied:
수정했습니다! 😄

kingbbode reviewed (changes requested):
피드백 확인 부탁드려요!

willing1026 committed:
# modify : 불필요한 컬럼명 입력 제거

kingbbode reviewed (approved):
깔끔하네요! 👍 
3단계까지 고생 많으셨어요 :)
이번 단계에서 마무리하실지, Advanced 미션까지 진행해보실지 DM 보내주세요!
